File: codeforces/1829/E/the_lakes.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> matrix;
    int m;
    int n;

    int dfs(const int i, const int j) {
        if (!(0 <= i && i < m) || !(0 <= j && j < n) || matrix[i][j] == 0) {
            return 0;
        }

        auto cur = matrix[i][j];
        matrix[i][j] = 0;
        
        int up = dfs(i+1, j); 
        int down = dfs(i-1, j);
        int left = dfs(i, j+1);
        int right = dfs(i, j-1);

        return cur + up + down + left + right;
    }

    int volOfBigestLake(vector<vector<int>> _matrix) {
        matrix = _matrix;
        m = matrix.size(),
        n = matrix[0].size();

        vector<int> ans = {0};

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] != 0) {
                    int lake_vol = dfs(i, j);
                    ans.push_back(lake_vol);
                }
            }
        }

        auto maxVol = *max_element(ans.begin(), ans.end());

        return maxVol;
    }
};

int main(int argc, char const *argv[])
{
    int ncases;
    cin >> ncases;
    
    vector<int> ans;

    while (ncases--) {
        Solution s;
        int m, n;
        cin >> m;
        cin >> n;

        vector<vector<int>> matrix(m, vector<int>(n, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                cin >> matrix[i][j];
            }
        }

        int volOfBigestLake = s.volOfBigestLake(matrix);
        ans.push_back(volOfBigestLake);
    }

    for (const int vol: ans) {
        cout << vol << endl;
    }
    
    return 0;
}

File: codeforces/1829/E/the_lakes.py
from typing import List
from dataclasses import dataclass

@dataclass
class Position:
    i: int
    j: int
    visited: bool = False

class Solution:
    def isAdj(self, pos1: Position, pos2: Position) -> bool:
        return (abs(pos1.i - pos2.i), abs(pos1.j - pos2.j)) in ((1, 0), (0, 1))

    def findLake(self) -> int:
        self.pos_with_water[0].visited = True
        q = [ self.pos_with_water[0] ]                
        vol = 0
        
        while len(q) > 0:
            cur_pos = q.pop()
            vol += self.matrix[cur_pos.i][cur_pos.j]
            
            for pos in filter(lambda pos: pos.visited == False and self.isAdj(pos, cur_pos), self.pos_with_water):
                pos.visited = True
                q.append(pos)
            
        return vol

    def volOfMaxLake(self, matrix: List[List[int]]) -> int:
        self.M, self.N = len(matrix), len(matrix[0])
        self.matrix = matrix
        self.pos_with_water: List[Position] = []
        ans = [0]

        for i in range(self.M):
            for j in range(self.N):
                if self.matrix[i][j] != 0:
                    self.pos_with_water.append(Position(i, j))

        while len(self.pos_with_water) > 0:
            ans.append(self.findLake())
            self.pos_with_water = list(filter(lambda pos: pos.visited == False, self.pos_with_water))
        
        return max(ans)

ncases = int(input())
ans = []

for _ in range(ncases):
    s = Solution()
    m, n = list(map(int, input().split(' ')))
    matrix = []

    for i in range(m):
        matrix.append(list(map(int, input().split(' '))))

    maxvol = s.volOfMaxLake(matrix)
    ans.append(maxvol)

for vol in ans:
    print(vol)
